### PR TITLE
fail early if web app URI is not reachable

### PR DESCRIPTION
**Pull request by vladak**
#3517 inspired me to add this check. Wanted to do this outside of argument parsing - if there is any problem with options this should fail first. These methods come from authnz plugins so this is code duplication however plugins needs to be independent.

### CODE REVIEW

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/Indexer.java [hunk 1]
```diff
@@ -161,9 +163,36 @@ public static void main(String[] argv) {
 
         boolean createDict = false;
 
+        int CONNECT_TIMEOUT = 1000;  // in milliseconds
+
         try {
             argv = parseOptions(argv);
 
+            if (webappURI != null) {
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/Indexer.java:171]: I'd move this check inside a separate method inside the `HostUtil` class. I think the `main` method is already too long.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/Indexer.java [hunk 2]
```diff
@@ -161,9 +163,36 @@ public static void main(String[] argv) {
...
+                            connectWorks = true;
+                            break;
+                        }
+                    }
+                } catch (URISyntaxException e) {
+                    LOGGER.log(Level.WARNING, String.format("URI not valid: %s", webappURI), e);
+                }
+
+                if (!connectWorks) {
+                    LOGGER.log(Level.SEVERE, webappURI + " is not reachable.");
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/Indexer.java:191]: It seems we're printing the message to `System.err` right before the exit in other places. Will the log be always displayed?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HostUtil.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        return true;
+    }
+
+    public static boolean isReachable(String webappURI, int timeOutMillis) {
+        boolean connectWorks = false;
+        try {
+            for (InetAddress addr : InetAddress.getAllByName(HostUtil.urlToHostname(webappURI))) {
+                int port = HostUtil.urlToPort(webappURI);
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/util/HostUtil.java:88]: The port can actually be checked only once before the `for`, right?

Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/util/HostUtil.java]: true !

**Review by ahornace — APPROVED** [review]

**Comment by vladak** [comment]
thanks @ahornace !

**Pull request merged**